rasa/engine/training/fingerprinting.py

```diff
@@ -0,0 +1,35 @@
+import logging
+from typing import Any, Dict, Text
+from typing_extensions import Protocol, runtime_checkable
+
+import rasa.shared.utils.io
+
+logger = logging.getLogger(__name__)
+
+
+@runtime_checkable
+class Fingerprintable(Protocol):
+    """Interface that enforces training data can be fingerprinted."""
+
+    def fingerprint(self) -> Text:
+        """Returns a unique stable fingerprint of the data."""
+        ...
+
+
+def calculate_fingerprint_key(
+    node_name: Text, config: Dict[Text, Any], inputs: Dict[Text, Fingerprintable]
+) -> Text:
+    """Calculates a fingerprint key that uniquely represents a single graph node run."""
+    fingerprint_data = {
+        "node_name": node_name,
```

Comment:
Should this be the component class?

Reply:
That's what we actually specified in the implementation proposal after having Fede's feedback